### PR TITLE
[java] LoggerIsNotStaticFinal and ProperLogger - make class-name configurable

### PR DESCRIPTION
**Pull request by bedla**
Hi, I am missing possibility to configure logger class name with property. What do you think about this PR? Thx, Ivos

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">This changeset introduces 0 new violations, 0 new errors and 0 new configuration errors,<br />
removes 0 violations, 0 errors and 2 configuration errors.<br />
<a href="http://chunk.io/pmd/8cceca3e4fb94a6cb486c50e21d96eed/diff/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by jsotuyod** [comment]
@bedla thanks for the PR! This certainly got me thinking…

Firstly, I'm unsure we actually need these 2 rules to be separate… and I also think the rules need some revamp:

1. I'd love to see them use type resolution instead of image matching
2. With type resolution, I doubt we would even care about the field name at all…

Do you have a scenario where this would not be true?

**Comment by bedla** [comment]
Thx, for response...

- I have no problem with mergin those two rules into one. But what are backward compatibility rules for such case? Create new rule with new ID and somehow deprecate old ones?
- I can investigate more type resolution instead of image matching. 
- Maybe I dont get it, but I think that field name matching is one of the advantage for those rules. Not having mixture of LOG, log, LOGGER, logger static fields, right?

**Comment by jsotuyod** [comment]
@bedla you are right about the field name.

As for type resolution, you can look at the docs [here](https://pmd.github.io/pmd-6.13.0/pmd_userdocs_extending_writing_pmd_rules.html#inside-an-xpath-query) on how to use it from XPath.

As for the merging, it depends… on this case I'd:

1. deprecate `LoggerIsNotStaticFinal`
2. add a new property `requireStaticFinalLogger` defaulting to `false` to `ProperLogger`, and use it to determine wether to even check for those, this way the rule's behavior remains unchanged, but people wanting `LoggerIsNotStaticFinal` can set it to `true`.
3. in PMD 7.0.0 we should change the default value to `true` (as it's a far more common approach)

**Comment by bedla** [comment]
cool, i will update PR this way

**Comment by adangel** [comment]
So this would mean, that [ProperLogger](https://pmd.github.io/pmd-6.13.0/pmd_rules_java_errorprone.html#properlogger) already covers the cases that [LoggerIsNotStaticFinal](https://pmd.github.io/pmd-6.13.0/pmd_rules_java_errorprone.html#loggerisnotstaticfinal) is detecting. One difference seems to be, that ProperLogger was implemented to be used with apache commons logging (class image is "Log"), while LoggerIsNotStaticFinal is targeting java.util.logging (class image is "Logger").

Looking at the test cases yields, that ProperLogger only looks at fields, that are already private+static+final. It would not complain, if you declare your logger public...

-> Conclusion: Yes, we need this new property requireStaticFinalLogger.

I'm wondering, whether ProperLogger works out of the box for java.util.logging? java.util.logging uses a String as the logger name, while apache commons logging uses the class (would additionally support String as well).

Example:
```
private static final java.util.logging.Logger LOG = Logger.getLogger(Foo.class.getName());
private static final org.apache.commons.logging.Log LOG = LogFactory.getLog(Foo.class);
private static final org.slf4j.Logger LOG = LoggerFactory.getLogger(Foo.class);
```

(I've added a slf4j example as well).

If we deprecate LoggerIsNotStaticFinal, we should make sure, that ProperLogging works with java.util.logging as well, so that it can be used as a replacement.

As for this PR, I'd suggest the following approach:
1. Get it working with typeresolution
2. Make sure the class type can be configured
3. Combine the rules (in the simplest case, you use the "|" (union) operator to concatenate the two xpath expressions)
4. Introduce the new property to only use the second part of the expression, if the property is true.

**Comment by bedla** [comment]
I have force-pushed update...
I not sure about rename of "staticLoggerName" to "loggerName" ... it is not static anymore (and I did not find how to deprecate property or keep it backward compatible without makeing XPath too complicated). I have added more tests. But I think that rule become backward incompatible. 
What do you think?

File: pmd-java/src/main/resources/category/java/errorprone.xml [hunk 1]
```diff
@@ -2633,23 +2636,36 @@ with the restriction that the logger needs to be passed into the constructor.
...
- not(FieldDeclaration[@Final='true'][@Static='true'][@Private='true'][.//VariableDeclaratorId[@Image=$staticLoggerName]]
- and
- //ArgumentList//ClassOrInterfaceType[@Image = ancestor::ClassOrInterfaceDeclaration/@Image or @Image = ancestor::EnumDeclaration/@Image])
- and
- not(FieldDeclaration[@Final='true'][@Private='true'][.//VariableDeclaratorId[@Image='log']]
- [count(.//VariableInitializer)=0]
- [ancestor::ClassOrInterfaceBody//StatementExpression[.//PrimaryExpression/descendant::*[@Image='log']][count(.//AllocationExpression)=0]]
- )]
+//ClassOrInterfaceBodyDeclaration[
+ FieldDeclaration//ClassOrInterfaceType[pmd-java:typeIs($loggerClass)]
```
[the remark this answers is not in the document]
Comment (bedla) [reply pmd-java/src/main/resources/category/java/errorprone.xml]: i see, fixed

File: pmd-java/src/main/resources/category/java/errorprone.xml [hunk 2]
```diff
@@ -2633,23 +2636,36 @@ with the restriction that the logger needs to be passed into the constructor.
...
+  and not (
+   (: check logger name :)
+   (
+    FieldDeclaration[$requireStaticFinalLogger][@Final=true()][@Static=true()][.//VariableDeclaratorId[@Image=$loggerName]]
```
[the remark this answers is not in the document]
Comment (bedla) [reply pmd-java/src/main/resources/category/java/errorprone.xml]: fixed

File: pmd-java/src/main/resources/category/java/errorprone.xml [hunk 3]
```diff
@@ -2633,23 +2636,36 @@ with the restriction that the logger needs to be passed into the constructor.
...
+   (: final logger initialized inside constructor :)
+   FieldDeclaration[@Final=true()][@Private=true()][.//VariableDeclaratorId[@Image=$loggerName]]
+   [count(.//VariableInitializer)=0]
+   [ancestor::ClassOrInterfaceBody//StatementExpression[.//PrimaryExpression/descendant::*[@Image=$loggerName]][count(.//AllocationExpression)=0]]
+  )
+]
 ]]>
                 </value>
             </property>
-            <property name="staticLoggerName" type="String" description="Name of the static Logger variable" value="LOG"/>
```
[the remark this answers is not in the document]
Comment (bedla) [reply pmd-java/src/main/resources/category/java/errorprone.xml]: fixed

**Comment by bedla** [comment]
I have updated PR

File: docs/pages/pmd/rules/java/errorprone.md [hunk 4]
```diff
@@ -2850,16 +2854,26 @@ with the restriction that the logger needs to be passed into the constructor.
...
+   )
+   and
+   (: in place initialization with method argument pointing to current class (even for Enums) :)
+   .//ArgumentList//ClassOrInterfaceType[@Image = ancestor::ClassOrInterfaceDeclaration/@Image or @Image = ancestor::EnumDeclaration/@Image]
+  )
+  and not (
+   (: final logger initialized inside constructor :)
+   FieldDeclaration[@Final=true()][@Private=true()][.//VariableDeclaratorId[@Image=$loggerName or @Image=$staticLoggerName]]
+   [count(.//VariableInitializer)=0]
+   [ancestor::ClassOrInterfaceBody//StatementExpression[.//PrimaryExpression/descendant::*[@Image=$loggerName or @Image=$staticLoggerName]][count(.//AllocationExpression)=0]]
```
Comment (bedla) [inline docs/pages/pmd/rules/java/errorprone.md]: I have a question about this `[count(.//AllocationExpression)=0]` part, when I remove it current tests are all green. What is it used for? Why creation new instance of something is considered as bug (there is not a test case for it)?
And I found another use-case that could be checked
```java
public class Xxx {
    private final Log log;
    Xxx() {
        this.log = MyLoggerFactory.getLogger(Xxx.class);
    }
}
  
public class Xxx {
    private final Log log;
    Xxx() {
        this.log = MyLoggerFactory.getLogger(Bar.class);
    }
}
```

Comment (adangel) [reply docs/pages/pmd/rules/java/errorprone.md]: I guess, the original author wanted to prevent things like this:

`private final Log log = new Log(MyClass.class)`

Usually, the logging framework provide factories, so that the logger instances are shared singletons and under control by the logging framework. With explicitly calling a constructor, we would not use the logging framework correctly (hopefully, the constructor is anyway not public, so this would be theoretical problem). And yes, if no test is failing, we have no test case for this.

Your other use case: Is the problem, you see here, that the logger for Bar is used in the class Xxx? Meaning, you first snippet is the correct one (no violation), while we should flag the second snippet? Or is the issue with initializing the field in the constructor? Actually, I would prefer to initialize the (final) field directly in the declaration....

File: pmd-java/src/main/resources/category/java/errorprone.xml [hunk 5]
```diff
@@ -190,7 +190,7 @@ for (int i = 0; i < 10; i++) {
         <description>
 The method Object.finalize() is called by the garbage collector on an object when garbage collection determines
 that there are no more references to the object. It should not be invoked by application logic.
-
```
Comment (adangel) [inline pmd-java/src/main/resources/category/java/errorprone.xml]: Please try to avoid changing unrelated whitespaces. These changes are only distracting. If they are necessary, then they should be in a separate PR only dealing with whitepaces.

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/errorprone/xml/ProperLogger.xml [hunk 6]
```diff
@@ -29,6 +29,7 @@ public class Foo {
         <description><![CDATA[
 Ok, special case
         ]]></description>
+        <rule-property name="loggerName">log</rule-property>
```
Comment (adangel) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/errorprone/xml/ProperLogger.xml]: This change is critical - the need to change/set the property means, the rule behaves differently, which means, we have a backwards compatibility problem....

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/errorprone/xml/ProperLogger.xml [hunk 7]
```diff
@@ -87,6 +89,99 @@ bug 1626232, extra loggers with different names are not allowed by default (use
...
+    public static final Log LOG = LogFactory.getLog(Foo.class);
+}
+ ]]>    </code>
+    </test-code>
+    <test-code>
+        <description><![CDATA[
+Public logger when static final required
+        ]]></description>
+        <rule-property name="requireStaticFinalLogger">true</rule-property>
+        <expected-problems>0</expected-problems>
```
Comment (adangel) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/errorprone/xml/ProperLogger.xml]: shouldn't we raise here a problem, since the logger is public and not private?

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/errorprone/xml/ProperLogger.xml [hunk 8]
```diff
@@ -87,6 +89,99 @@ bug 1626232, extra loggers with different names are not allowed by default (use
...
+public class Foo {
+    private static final MyLog INVALID_LOG_NAME = LogFactory.getLog(Foo.class);
+}
+ ]]>    </code>
+    </test-code>
+    <test-code>
+        <description><![CDATA[
+Public logger
+        ]]></description>
+        <expected-problems>0</expected-problems>
```
Comment (adangel) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/errorprone/xml/ProperLogger.xml]: This case was flagged by the old rule implementation - so public loggers should not be allowed in any case...

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/errorprone/xml/ProperLogger.xml [hunk 9]
```diff
@@ -87,6 +89,99 @@ bug 1626232, extra loggers with different names are not allowed by default (use
...
+ ]]>    </code>
+    </test-code>
+    <test-code>
+        <description><![CDATA[
+Logger initialized as String constant from class name
+        ]]></description>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+public class Foo {
+    Log LOG = LogFactory.getLog(Foo.class.getName());
```
Comment (adangel) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/errorprone/xml/ProperLogger.xml]: Since the field is not private, the rule should should flag this... the test case should be `private static final Log LOG = LogFactory.getLog(Foo.class.getName());`.

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/errorprone/xml/ProperLogger.xml (hunk 7 above)
Comment (adangel) [reply pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/errorprone/xml/ProperLogger.xml]: It seems, my earlier statement

> Looking at the test cases yields, that ProperLogger only looks at fields, that are already 
> private+static+final. It would not complain, if you declare your logger public...

was wrong: Yes, there were no test cases for public loggers, but the rule still raised a violation...
